### PR TITLE
add arrow_block support for with block of type bytes

### PR DESCRIPTION
**Pull request by wybryan**
…type

1. arrow_block support for bytes as data type in addition to
paarrow.Table

2. add test case for it to showcase the usage

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wybryan** [comment]
Hi there, 
I would like to contribute to this wonderful project. 

The feature I am making is to support the data conversion from raydp Spark dataframe -> ray Dataset -> pandas dataframe.

Currently, it will crush as the line "assert isinstance(block, pyarrow.Table), block" of the function add_block() defined in arraow_block.py. The reason is because this function asumes  the input data type of the block to be pyarrow.Table, which is not the case where the ray Dataset is created from raydp Spark dataframe conversion, which is of data type as bytes.

Therefore, to support this additional block type, I add a type theck and conversion from bytes to Paarrow.Table in add_block(), together with a unit test to showcase its usage.

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
LINT: 



python/ray/data/tests/test_raydp_dataset.py:52:15: Q000 Remove bad quotes

**Comment by wybryan** [comment]
@ericl I have fixed the lint issue.

cheers. Bryan

**Comment by ericl** [comment]
There seems to be a merge conflict from master.

**Comment by wybryan** [comment]
@ericl Yes, I can see arrow_block.py has been modified quite heavily because of https://github.com/ray-project/ray/pull/20721.

does #20721 PR solve the issue I am trying to fix? @kfstorm

**Comment by kfstorm** [comment]
@wybryan I don't think so. That PR only does refactoring. But I'm actively working on bringing pandas DataFrame block format to Ray Dataset. Not sure if it's related to your work. Check #20719

**Comment by wybryan** [comment]
@ericl @kfstorm Hi guys, I managed to fix this issue with the latest code.

What is it: 
To fix code crush where Dataset is constructed with block of type bytes instead pyarrow.Table.

Test: see the corresponding unit test. It is a common use case where Dataset is the result of conversion from RayDP Spark Dataframe.

File: python/ray/data/impl/table_block.py [hunk 1]
```diff
@@ -129,6 +129,9 @@ class TableBlockAccessor(BlockAccessor):
     def __init__(self, table: Any):
         self._table = table
 
+    def get_block_data(self) -> Block:
```
Comment (kfstorm) [inline python/ray/data/impl/table_block.py:132]: How about just `get_block` or a property `block`?

**Review by ericl — CHANGES_REQUESTED** [review]
I don't think the design of this is right. If we want to support bytes, we can convert that to pyarrow.Table within the `from_spark()` call with zero copy using pyarrow APIs, then I don't think we need to break the abstractions here.

File: python/ray/data/impl/table_block.py (hunk 1 above)
Comment (wybryan) [reply python/ray/data/impl/table_block.py:132]: @kfstorm I can do that, I'll change function name to get_block().

**Comment by wybryan** [comment]
@ericl I am not sure I can agree, as it is absolutely nothing to do with from_spark(). Let me explain this more in detail.

1. Ray Dataset is Ray's native data type. Any data fullfills the definition of Ray Dataset is to be supported directly by Ray. This is nothing to do with external system such as raydp.from_spark(). I use this in unit test is just to showcase what is currently missing in Ray's codebase.

2.  So what is the definition of a Dataset defined by Ray? Ray defines that the Dataset can be generated from 3 ways: plain pyhton list, pyarrow.Table, bytes. And this definiton is defined inside Ray across the board, which is evidently written in the current codebase before this PR, let's see what they are:

2.1 to support Dataset generated from list (i.e., the block type is list), there is a SimpleBlockBuilder and SimpleBlockAccessor, they are 1:1 mapping between the two, therefore in the factory method build(...), an accessor of type SimpleBlockAccessor can only and will only build a block of type SimpleBlockBuilder.

2.2 to support Dataset generated from pyarrow.Table & bytes ( (i.e., the block type is pyarrow.Table or bytes). There is ArrowBlockAccessor and ArrowBlockBuilder. This abstraction is already in the codebase before this PR, but there is "bug" or overlook of this design, that's what this PR is about, this PR is to fix this missing piece. Let's zoom what problem it is here:

2.2.1 when the Dataset is of type pyarrow.Table, everyhing works., but not with bytes. The problem is here the ArrowBlockAccessor is no longer 1:1 mapping to ArrowBlockBuilder, it is 2:1 mapping relationship. The ArrowBlockAccessor can be constructed either with pyarrow.Table, or with bytes, where there is an automatic conversion done within ArrowBlockAccessor class as shown below in the codebase:

    def from_bytes(cls, data: bytes):
        reader = pyarrow.ipc.open_stream(data)
        return cls(reader.read_all())

It means ArrowBlockerAccessor can be constructed by 2 different data types, but this information gets lost when we build ArrowBlockBuilder() with hardcoded _block_type = "pyarrow.Table" as shown below:

    def __init__(self):
        if pyarrow is None:
            raise ImportError("Run `pip install pyarrow` for Arrow support")
        TableBlockBuilder.__init__(self, pyarrow.Table)    ------  HARDCODED


How to solve it? Well, I believe the abstraction design is solid, as it already has all information we need, we just need to NOT HARDCODE when we construct ArrowBlockBuilder, in this way, when ArrowBlockAccessor is constructed, it knows what block type it should be using when building the corresponding ArrowBlockBuilder accordingly, as a result, we have consistent ArrowBlockAccessor & ArrowBlockBuilder aligned with the internal block type, job done. 

Therefore, I would argue this PR is not to break abstraction design, but rather fix the flaw the codebase currently have. It is nothing to do with from_spark(), which is just easier for me to generate a legit Dataset to demonstrate this PR. In addition, the dataset generated by from_spark() can work with to_arrow(), to_pandas_ref(), to_dask(), to_modin() along with native dataset API such as ds.mean("A") etc., which further indicates there is inconsistency in current codebase when dealing with dataset.

**Comment by ericl** [comment]
But isn't there a simpler solution? If you changed ArrowBlockBuilder to auto-convert bytes to pyarrow.Table on `add_block`, none of this tracking logic is needed.

**Comment by wybryan** [comment]
@ericl Hi Eric

what you suggested is indeed a simpler solution, my concern is the abstraction at moment intended to hide explicit type conversion, do we want to have special handling of daat type just for this case in the add_block? 

So perhaps @kfstorm you can comment on this as well, after all the abstraction may affect what you are doing as well: whether we do explicit conversion for block type of bytes specically in the add_block, or we accept the current proposal where the solution may be slightly less staightforward, but consistent for other block types.

**Comment by ericl** [comment]
So one other consideration here is that we might be able to remove the need
for a bytes type in the future (there's a separate thread where we can
support Arrow serialization natively in Ray, which would render the need
for bytes type moot). Given this, I'd prefer we opt for something simple we
can remove in the future once a better way is available.

**Comment by kfstorm** [comment]
@wybryan I think I get your point about this code change. But I'm still not sure what's the use case. Could you add a test case without involving raydp?

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Pull request closed without merging**